crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -0,0 +1,408 @@
+        /// per_module[slot_state[i].module][slot_state[i].per_module_index]
+        /// == i.
+        per_module_index: usize,
+    },
+}
+
+impl SlotState {
+    /// Get the index of this slot in the global free list.
+    fn free_list_index(&self) -> usize {
+        match self {
+            &Self::Empty { free_list_index }
+            | &Self::Affinity {
+                free_list_index, ..
+            } => free_list_index,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Update the index of this slot in the global free list.
+    fn update_free_list_index(&mut self, index: usize) {
+        match self {
+            &mut Self::Empty {
+                ref mut free_list_index,
+            }
+            | &mut Self::Affinity {
+                ref mut free_list_index,
+                ..
+            } => {
+                *free_list_index = index;
+            }
+            _ => panic!("Taken in free list"),
+        }
+    }
+
+    /// Get the index of this slot in its per-module free list.
+    fn per_module_index(&self) -> usize {
+        match self {
+            &Self::Affinity {
+                per_module_index, ..
+            } => per_module_index,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Update the index of this slot in its per-module free list.
+    fn update_per_module_index(&mut self, index: usize) {
+        match self {
+            &mut Self::Affinity {
+                ref mut per_module_index,
+                ..
+            } => {
+                *per_module_index = index;
+            }
+            _ => panic!("Taken in per-module free list"),
+        }
+    }
+}
+
+impl PoolingAllocationState {
+    /// Create the default state for this strategy.
+    pub(crate) fn new(strategy: PoolingAllocationStrategy, max_instances: usize) -> Self {
+        let ids = (0..max_instances).collect::<Vec<_>>();
+        match strategy {
+            PoolingAllocationStrategy::NextAvailable => PoolingAllocationState::NextAvailable(ids),
+            PoolingAllocationStrategy::Random => PoolingAllocationState::Random(ids),
+            PoolingAllocationStrategy::ReuseAffinity => PoolingAllocationState::ReuseAffinity {
+                free_list: ids,
+                per_module: HashMap::new(),
+                slot_state: (0..max_instances)
+                    .map(|i| SlotState::Empty { free_list_index: i })
+                    .collect(),
+            },
+        }
+    }
+
+    /// Are any slots left, or is this allocator empty?
+    pub(crate) fn is_empty(&self) -> bool {
+        match self {
+            &PoolingAllocationState::NextAvailable(ref free_list)
+            | &PoolingAllocationState::Random(ref free_list) => free_list.is_empty(),
+            &PoolingAllocationState::ReuseAffinity { ref free_list, .. } => free_list.is_empty(),
+        }
+    }
+
+    /// Internal: remove a slot-index from the global free list.
+    fn remove_free_list_item(
```

Comment:
Nitpick: maybe name this `remove_global_free_list_item` to clarify the difference between this and `remove_module_free_list_item`?

Ditto about freeling like a free function.

Reply:
Yup, done on both.